File: guests/js/table.ts
// The table scene: a header click is a REQUEST, and the platform sorts none.
//     KAYA_SELFTEST=table node guests/js/table.ts

import * as kaya from "kaya-gui";

const Item = kaya.record({ name: String, size: String }, "Item");
type ItemFields = kaya.Fields<typeof Item.schema>;

const app = new kaya.App();

// The guest's sort policy; the platform never has one.
let sorted: [number, boolean] | null = null;

function onSort(column: number): void {
  const current = sorted;
  const descending = current !== null && current[0] === column && !current[1];
  sorted = [column, descending];
  const entries = items.items();
  const keyOf = column === 0 ? (e: [kaya.Key, ItemFields]) => e[1].name : (e: [kaya.Key, ItemFields]) => e[1].size;
  entries.sort((a, b) => {
    const x = keyOf(a);
    const y = keyOf(b);
    const order = x < y ? -1 : x > y ? 1 : 0;
    return descending ? -order : order;
  });
  // Each key to the end, in the target order.
  for (const [key] of entries) {
    items.moveToEnd(key);
  }
  const indicator = descending ? kaya.Sort.desc(column) : kaya.Sort.asc(column);
  items.setColumns(["Name", "Size"], { sort: indicator });
}

let items!: kaya.Collection<ItemFields, kaya.Row<typeof Item.schema>>;

app.window(() => {
  items = kaya.collection(Item);
  // The root is a row: the For's container is the only column.
  kaya.row(() => {
    // Grown on purpose: ungrown, a table hugs its rows.
    for (const item of items.columns(["Name", "Size"], { onSort, grow: 1 })) {
      kaya.row(() => {
        kaya.label({ bind: item.name });
        kaya.label({ bind: item.size });
      });
    }
  });
  for (const [key, name, size] of [
    ["b", "banana", "30"],
    ["a", "apple", "10"],
    ["c", "cherry", "20"],
  ] as const) {
    items.insert(key, Item({ name, size }));
  }
});

app.run();

File: guests/haskell/table.hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE TypeApplications #-}

-- The table scene, Haskell port — guests/rust/table.rs,
-- tools/scenes/table.steps.

import Data.IORef (newIORef, readIORef, writeIORef)
import Data.List (sortBy)
import Data.Ord (comparing)
import Data.Proxy (Proxy (..))
import GHC.Generics (Generic)

import KayaApp
import KayaWire (Value (..))

data TableItem = TableItem {name :: String, size :: String} deriving (Generic)

instance KayaRecord TableItem

main :: IO ()
main = kayaMain $ \app -> do
  -- The guest's sort policy; the platform never has one.
  sorted <- newIORef (Nothing :: Maybe (Int, Bool))
  (items, table) <- buildTx app $ do
    items <- collectionOf (Proxy :: Proxy TableItem)
    -- The root is a row so the For's container is the scene's only
    -- column-kind widget (the reorder scene's rule).
    (table, _) <-
      forEach (recordHandle items) $
        rowOf
          [ label (field @"name" @TableItem),
            label (field @"size" @TableItem)
          ]
    -- Grown on purpose: ungrown would hug its rows (docs/tables-plan.md
    -- decision 8).
    setGrow table 1
    columns table ["Name", "Size"] sortNone
    root <- row [pure table]
    mount root
    mapM_
      (\(k, n, s) -> insertRecord items (VStr k) (TableItem n s))
      [("b", "banana", "30"), ("a", "apple", "10"), ("c", "cherry", "20")]
    return (items, table)
  onSort app table $ \column -> do
    current <- readIORef sorted
    let desc = case current of
          Just (c, d) -> c == column && not d
          _ -> False
    writeIORef sorted (Just (column, desc))
    submitTx app $ do
      entries <- recordItems items
      let keyOf (_, item) = if column == 0 then name item else size item
          ordered0 = sortBy (comparing keyOf) entries
          ordered = if desc then reverse ordered0 else ordered0
      -- Keys, never indices.
      mapM_ (\(k, _) -> moveToEnd (recordHandle items) k) ordered
      columns table ["Name", "Size"] $
        if desc then sortDesc column else sortAsc column
